sesman/session.c

```diff
@@ -676,6 +679,27 @@ session_start_fork(tbus data, tui8 type, struct SCP_SESSION *s)
                 g_snprintf(text, 255, "%d", g_cfg->sess.kill_disconnected);
                 g_setenv("XRDP_SESMAN_KILL_DISCONNECTED", text, 1);
 
+                /* prepare the Xauthority stuff */
+                if (g_getenv("XAUTHORITY") != NULL)
+                {
+                    g_snprintf(authfile, 255, "%s", g_getenv("XAUTHORITY"));
+                }
+                else
+                {
+                    g_snprintf(authfile, 12, "%s", ".Xauthority");
+                }
+
+                /* Create the cookie */
+                for (i = 0; i < 32; i++)
+                {
+                    g_random((char *) &cookie_tmpval, 1);
+                    sprintf(&cookie[i], "%02X", cookie_tmpval & 0xff);
+                }
+                cookie[32] = '\0';
```

Comment:
Calling g_random() 32 times to fill 16 bytes is highly ineffective. Please call g_random() once to get 16 random bytes. Then use g_bytes_to_hexstr() to convert it to text.